File: Identite.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define T 2   // Taille des matrices
#define N 10  // Nombre de termes dans la somme

// Calcule la factorielle d'un entier
double factorial(int n) {
    if (n == 0) return 1.0;
    double result = 1.0;
    for (int i = 1; i <= n; i++)
        result *= i;
    return result;
}

// Multiplie deux matrices de taille T x T
void mat_mult(double A[T][T], double B[T][T], double result[T][T]) {
    double temp[T][T] = {0};
    for (int i = 0; i < T; i++) {
        for (int j = 0; j < T; j++) {
            for (int k = 0; k < T; k++) {
                temp[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    // Copie dans result
    for (int i = 0; i < T; i++)
        for (int j = 0; j < T; j++)
            result[i][j] = temp[i][j];
}

// Calcule A^n pour une matrice carrée A
void mat_power(double A[T][T], int power, double result[T][T]) {
    // Initialise result à l'identité
    for (int i = 0; i < T; i++)
        for (int j = 0; j < T; j++)
            result[i][j] = (i == j) ? 1.0 : 0.0;

    double temp[T][T];
    for (int p = 0; p < power; p++) {
        mat_mult(result, A, temp);
        for (int i = 0; i < T; i++)
            for (int j = 0; j < T; j++)
                result[i][j] = temp[i][j];
    }
}

// Fonction Identite
void Identite(double A[T][T], double Somme[T][T]) {
    // Initialise la somme à zéro
    for (int i = 0; i < T; i++)
        for (int j = 0; j < T; j++)
            Somme[i][j] = 0.0;

    double temp[T][T];

    for (int i = 0; i < N; i++) {
        double pi = factorial(i);
        for (int j = 0; j < N; j++) {
            double pj = factorial(j);
            double coeff = pow(-1.0, j) / (pi * pj);
            mat_power(A, i + j, temp);
            for (int x = 0; x < T; x++)
                for (int y = 0; y < T; y++)
                    Somme[x][y] += coeff * temp[x][y];
        }
    }
}

// Affiche une matrice
void print_matrix(double M[T][T]) {
    for (int i = 0; i < T; i++) {
        for (int j = 0; j < T; j++) {
            printf("%8.4f ", M[i][j]);
        }
        printf("\n");
    }
}

int main() {
    double A[T][T] = {
        {0, 1},
        {0.125, 0.25}
    };

    double S[T][T];
    Identite(A, S);

    printf("Resultat de la somme:\n");
    print_matrix(S);

    return 0;
}

File: Identite.f90
program identite_main
  implicit none
  integer, parameter :: T = 2, N = 10
  real(8) :: A(T, T), S(T, T)

  ! Exemple de matrice
  A = reshape((/ 0.0d0, 1.0d0, &
                 0.125d0, 0.25d0/), shape(A))

  call Identite(A, N, T, S)

  print *, 'Resultat de la somme :'
  call print_matrix(S)

contains

  function factorial(n) result(fact)
    integer, intent(in) :: n
    real(8) :: fact
    integer :: i
    fact = 1.0d0
    if (n == 0) return
    do i = 1, n
      fact = fact * dble(i)
    end do
  end function factorial

  subroutine mat_mult(A, B, result, T)
    integer, intent(in) :: T
    real(8), intent(in) :: A(T,T), B(T,T)
    real(8), intent(out) :: result(T,T)
    integer :: i, j, k
    result = 0.0d0
    do i = 1, T
      do j = 1, T
        do k = 1, T
          result(i,j) = result(i,j) + A(i,k) * B(k,j)
        end do
      end do
    end do
  end subroutine mat_mult

  subroutine mat_power(A, power, T, result)
    integer, intent(in) :: power, T
    real(8), intent(in) :: A(T,T)
    real(8), intent(out) :: result(T,T)
    real(8) :: temp(T,T), inter(T,T)
    integer :: i, j

    ! Initialise à l'identité
    result = 0.0d0
    do i = 1, T
      result(i,i) = 1.0d0
    end do

    do i = 1, power
      call mat_mult(result, A, inter, T)
      result = inter
    end do
  end subroutine mat_power

  subroutine Identite(A, N, T, Somme)
    integer, intent(in) :: N, T
    real(8), intent(in) :: A(T,T)
    real(8), intent(out) :: Somme(T,T)
    real(8) :: pi, pj, coeff, temp(T,T)
    integer :: i, j

    Somme = 0.0d0

    do i = 0, N - 1
      pi = factorial(i)
      do j = 0, N - 1
        pj = factorial(j)
        coeff = (-1.0d0)**j / (pi * pj)
        call mat_power(A, i + j, T, temp)
        Somme = Somme + coeff * temp
      end do
    end do
  end subroutine Identite

  subroutine print_matrix(M)
    real(8), intent(in) :: M(:,:)
    integer :: i, j, T
    T = size(M,1)
    do i = 1, T
      do j = 1, T
        write(*,'(F10.4)', advance='no') M(i,j)
      end do
      print *
    end do
  end subroutine print_matrix

end program identite_main
